[ClangImporter] Objective-C import decision remarks

beccadax opened the pull request:
Broadly, this PR can be divided into four parts:

1. **General `SwiftDeclConverter` refactoring.** This is trying to move towards a world where each `SwiftDeclConverter` instance is used to convert one and only one declaration, and where the reentrancy of `SwiftDeclConverter` is significantly better defined than it is today. Among other things, it ensures that `SwiftDeclConverter` is only entered from the outside through five specific entry points.

2. **Enum case import refactoring and laziness improvements.** This debrides several tangled code paths in `SwiftDeclConverter::VisitEnumDecl()` that previously were greedily importing the cases of most enums. Now, only clang enums that are actually imported as Swift enums (as opposed to those imported as OptionSet structs or global constants) have their cases imported immediately; the others are loaded lazily on demand. The code that remains is also better factored.

3. **`DiagnosticEngine` improvements.** `clang::NamedDecl*`s can now be used as arguments for diagnostics; `%select` is now compatible with all arguments (and does an "empty" or "null" check for ones that don't map to integers); and a new `-verify-additional-file` flag allows the diagnostic verifier to look for expectations in files outside of the module's actual source files, so that (for instance) we can check diagnostics emitted in header files.

4. **Objective-C import decision remarks.** The actual point of all of the above.

With import decision remarks enabled, ClangImporter will (try to) annotate either every clang declaration in a given module, or every declaration it fails to import, with remarks explaining:

* Whether it could import the declaration.
* If it could, what name(s) it was imported as.
* If it couldn't, why it couldn't. (This is the hard part—the reasons are currently pretty vague, occasionally misleading, and sometimes missing entirely.)

For example (using the Foundation.h from clang-importer-sdk in our test suite):

```
Foundation.h:161:33: remark: Objective-C method '-[NSDictionary count]' imported as getter 'Foundation.(file).NSDictionary._'
@property (readonly) NSUInteger count;
                                ^
Foundation.h:161:33: remark: Objective-C property 'NSDictionary::count' imported as property 'Foundation.(file).NSDictionary.count'
@property (readonly) NSUInteger count;
                                ^
Foundation.h:162:1: remark: Objective-C method '-[NSDictionary objectForKey:]' imported as instance method 'Foundation.(file).NSDictionary.object(forKey:)'
- (nullable ObjectType)objectForKey:(nonnull KeyType)aKey;
^
Foundation.h:93:48: remark: could not import Objective-C class interface 'NSEnumerator': Swift cannot import forward declarations
@class NSString, NSArray, NSDictionary, NSSet, NSEnumerator;
                                               ^
Foundation.h:163:1: remark: could not import Objective-C method '-[NSDictionary keyEnumerator]': it uses a type that can't be imported
- (nonnull NSEnumerator *)keyEnumerator;
^
```

-------

**To Do:**

- [x] Fully annotate Foundation.h with its import decision remarks and enable that part of the test
- [x] Fix lack of import remarks for subscripts 
- [ ] Fix forcing of categories
- [ ] Fix multiple emission of remarks for forward declarations
- [ ] Fix multiple emission of remarks for some accessors
- [ ] Fix lack of unavailable for `classForPortCoder` class method alternate
- [ ] Clean up the way we filter by module and make submodule filtering/forcing work correctly
- [ ] Add tests for `-verify-additional-file`
- [ ] Look for other interesting examples in the Swift test suite and test their import decision remarks, too

beccadax committed:
[ClangImporter] [NFC] Add method for caching decls

brentdax commented:
@swift-ci please test source compatibility

beccadax committed:
[NFC] Canonicalize in cacheImportedDecl()

brentdax commented:
(Looks okay, failures unrelated.)

beccadax committed:
[NFC] Minimize entry into SwiftDeclConverter

SwiftDeclConverter is now entered through only five static methods, separating it more cleanly from the rest of ClangImporter.

beccadax committed:
[NFC] [ClangImporter] Don’t force top-level case constants

For `EnumKind::Constants`, this was dead code, as `VisitEnumDecl` was returning before this point. For `EnumKind::Unknown`, this appears be unnecessary.

beccadax committed:
[NFC] Make dumpSwiftLookupTables() non-dead

Ensures it can be called in the debugger.

beccadax committed:
[NFC] [ClangImporter] Refactor canonical enum case

Extracts a helper type to encapsulate the table used to canonicalize cases, and simplifies the logic by largely merging the handling of unavailable cases with that of canonical cases.

NOTE: In theory, this could cause a behavior change if canonicalEnumConstants was ever used to look up the same case twice. (Previously, `unimported` would have been nullptr, which would cause it to go down the “case is an alias” path; now, it would effectively be the same as `canonical` used to be, which would not go down that path.) I don’t believe this ever actually happened, but I’m leaving a breadcrumb in this commit message in case a bug is ever bisected to this commit.

# Conflicts:
#	lib/ClangImporter/ImportDecl.cpp

beccadax committed:
[NFC] Extract logic to import enum constants

Makes VisitEnumDecl() a little more manageable.

# Conflicts:
#	lib/ClangImporter/ImportDecl.cpp

# Conflicts:
#	lib/ClangImporter/ImportDecl.cpp

beccadax committed:
[NFC] [ClangImporter] Disentangle OptionSet import

…from enum case importing. This allows both to be simplified significantly.

# Conflicts:
#	lib/ClangImporter/ImportDecl.cpp

beccadax committed:
[NFC] [ClangImporter] Lazy-load OptionSet members

# Conflicts:
#	lib/ClangImporter/ImportDecl.cpp

beccadax committed:
Begin emitting decision remarks from ClangImporter

This commit introduces partial support for “import decision remarks”. These describe whether, and hopefully eventually why, Swift chose to import a particular clang declaration.

In this commit, they are always on for all modules. There are no tests yet, and a couple of existing tests fail because the new output confuses them.

beccadax committed:
Prevent spurious ImportDecision diagnoses

beccadax committed:
Improve decl names used in ImportDecision remarks

beccadax committed:
Add flags to filter ImportRemarks

beccadax committed:
Remove versions, add availability to import decision remarks

beccadax committed:
[NFC] Add debug helper for selector matching

beccadax committed:
Start describing the reasons for import failures

beccadax committed:
Refine import decision remark messages

beccadax committed:
[WIP] Allow additional files for diagnostic verifier

This change adds a frontend flag, -verify-additional-file, which can be used to pass extra files directly to the diagnostic verifier. These files are not otherwise considered to be Swift source files; they are not compiled or even properly parsed.

This feature can be used to verify diagnostics emitted in non-source files, such as in module interfaces or header files.

[WIP because of missing tests.]

beccadax committed:
Test import decision remarks

brentdax commented:
@swift-ci please test source compatibility

beccadax committed:
Try to import everything before emitting remarks

beccadax committed:
[ClangImporter] Note overlay declaration uses

brentdax commented:
(Second round of source compat suite runs, covering the ClangImporter refactoring, all look good.)

beccadax committed:
[NFC] Generalize %select in diagnostics

Now every diagnostic kind supports %select (and %s, although it’s not necessarily very useful for many). For enums, it selects enum cases by their raw values. For pointer types, it works like `%select{null text|not null text}`. For other stuctured data (strings, versions, etc.) it means `%select{empty text|not empty text}`.

beccadax committed:
[ClangImporter] Note details in import remarks

Replaces the `ImportReason` enum with an array of `ImportRemark::Note` structs, each containing a diagnostic to emit. This allows import decisions to be much more detailed, including declaration names and kinds, and will help make it possible to thread decisions much deeper into ClangImporter.

# Conflicts:
#	include/swift/AST/DiagnosticsClangImporter.def

beccadax committed:
[NFC] Allow clang::NamedDecls in diagnostics

beccadax committed:
Emit import decision remarks for subscripts

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
